content/ADR-200-raycast-sdk-component.md

```diff
@@ -0,0 +1,209 @@
+
+The origin point and global direction are calculated as follow:
+
+```typescript
+// first calculate the global origin of the raycast
+const globalOrigin = entity.worldMatrix.multiply(vec4(raycast.originOffset.xyz, 1))
+
+// and then calculate the global direction, relative to the
+let globalDirection = Vector3.Forward()
+if (raycast.globalDirection) {
+  // this is the simplest one, for example Vector3.Down() to evaluate if
+  // there is a floor and how far it is. No matter the local rotation, tilt or yaw
+  globalDirection = raycast.globalDirection
+} else if (raycast.localDirection) {
+  // then localDirection, is used to detect collisions in a path
+  // i.e. Vector3.Forward(), it takes into consideration the rotation of
+  // the entity to perform the raycast in local coordinates
+
+  const globalTarget = entity.worldMatrix.multiply(vec4(raycast.localDirection.xyz, 1)) as Vector3
+  globalDirection = globalTarget.subtract(globalOrigin)
+} else if (raycast.globalTarget) {
+  // this one is to make it easy to point towards a pin-pointed element
+  // in global space, like a fixed tower
+  globalDirection = raycast.globalTarget.subtract(globalOrigin)
+}
+```
+
+### How much times a raycast is executed?
+
+That depends entirely on the `bool continuous` property. If set to `true`, the raycast MUST execute on every scene tick. Otherwise, the raycast MUST execute within the scene tick they were added.
+
+In an ideal scenario in which all messages and raycast are processed within the time window/quota defined in (ADR-148)[/adr/ADR-148], the scenes will receive the raycast result on the next update loop of the ECS. Enabling immediate-mode raycasts like the follwing example:
+
+```typescript
+function laserDamageSystem() {
+  for (const [entity, _turret] of engine.entitiesWith(LaserTurret)) {
+    const result = raycast(entity, Vector3.Forward())
+    if (result?.hits.length) {
+      // apply damage to all hitted entities
+    }
+  }
+}
+
+// this helper creates sets a Raycast component in the entity and return
+// its RaycastResult
+function raycast(entity: Entity, direction: Vector3) {
+  Object.assign(Raycast.getMutableOrCreate(entity), {
+    // relative to the entity's rotation
+    localDirection: direction,
+    continous: true,
+    queryType: ALL,
+  })
+  return RaycastResult.getOrNull(entity)
+}
+```
+
+### Usage of the `timestamp` property
+
+The timestamp property is a correlation number, only defined by the scene. The renderer MUST copy the value of the `timestamp` from the Raycast component to the RaycastResult component.
+
+### Performance considerations
+
+Having multiple continuous raycasts in a scene may have a big performance penalty. As a scene developer, it is RECOMMENDED to keep this number at minimum. As a renderer implementator, the compute of raycasts MUST count towards the execution quota of each scene to prevent a non-optimized scene from affecting the overall quality of the experience.
```

Comment:
What if we move this section out of semantics and move it next to the RFC sections as a `##` header?